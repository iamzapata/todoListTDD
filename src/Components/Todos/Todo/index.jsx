import React, { PureComponent } from "react"
import { shape, string, number, func } from "prop-types"

class Todo extends PureComponent {
  render() {
    const {
      todo: { completed, text, id },
      toggleTodo
    } = this.props
    return (
      <span
        className="Todo"
        onChange={() => {
          toggleTodo(id)
        }}
        style={{ textDecoration: completed ? "line-through" : "" }}
      >
        {text}
        <input type="checkbox" />
      </span>
    )
  }
}

Todo.propTypes = {
  todo: shape({
    id: number,
    text: string
  }).isRequired,
  toggleTodo: func.isRequired
}

export default Todo
