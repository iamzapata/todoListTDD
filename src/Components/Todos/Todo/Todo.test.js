import React from "react"
import { render, cleanup, fireEvent } from "react-testing-library"
import Todo from "./index"
import { type } from "ramda"
import "jest-dom/extend-expect"

describe("<Todo />", () => {
  afterEach(cleanup)

  const createProps = ({ completed = false } = {}) => ({
    todo: {
      id: 1,
      text: "Learn TypeScript",
      completed
    },
    toggleTodo: jest.fn()
  })

  it("Should render self", () => {
    const { container } = render(<Todo {...createProps()} />)

    expect(container.firstChild).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })

  it("Should have Todo className", () => {
    const { container } = render(<Todo {...createProps()} />)
    const classNameSelection = container.querySelectorAll(".Todo")
    expect(classNameSelection.length).toEqual(1)
  })

  it("Should have a wrapping span", () => {
    const { container } = render(<Todo {...createProps()} />)

    expect(type(container.firstChild)).toBe("HTMLSpanElement")
  })

  it("Should display the todo's text", () => {
    const { queryByText } = render(<Todo {...createProps()} />)
    const todo = queryByText("Learn TypeScript")

    expect(todo).toBeTruthy()
  })

  it("Should have a checkbox input", () => {
    const { container } = render(<Todo {...createProps()} />)
    const checkbox = container.querySelector("input[type=checkbox]")

    expect(checkbox).toBeTruthy()
  })

  it("Should show a line strike-through when todo is completed", () => {
    const { queryByText } = render(
      <Todo {...createProps({ completed: true })} />
    )
    const todo = queryByText("Learn TypeScript")

    expect(todo).toHaveStyle("text-decoration: line-through")
  })

  it("Should not show a line strike-through when todo is completed", () => {
    const { queryByText } = render(<Todo {...createProps()} />)
    const todo = queryByText("Learn TypeScript")

    expect(todo).not.toHaveStyle("text-decoration: line-through")
  })

  it("Should call toggleTodo when checkbox clicked/unclicked", () => {
    const props = createProps()
    const { container } = render(<Todo {...props} />)

    const checkbox = container.querySelector("input[type=checkbox]")

    checkbox.checked = true
    fireEvent.change(checkbox)
    checkbox.checked = false
    fireEvent.change(checkbox)

    expect(props.toggleTodo).toHaveBeenCalledTimes(2)
    expect(props.toggleTodo).toHaveBeenNthCalledWith(1, props.todo.id)
    expect(props.toggleTodo).toHaveBeenNthCalledWith(2, props.todo.id)
  })
})
