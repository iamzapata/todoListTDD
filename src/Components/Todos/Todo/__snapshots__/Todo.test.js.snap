// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> Should render self 1`] = `
<span
  class="Todo"
>
  Learn TypeScript
  <input
    type="checkbox"
  />
</span>
`;
